### DISCUSSION TITLE
Liquidsoap and Stereo Tool on a RPI3B+ : Optimization, tips, advices...

### DISCUSSION THREAD

**Discussion by nordseele** [savonet/liquidsoap#1782] [Q&A] [answered]
Hi, 
I'm trying to process the output of Liquid Soap with Stereo Tool on a Raspberry Pi 3B+ 

I've installed the 32bits cmd line version of Stereo Tool and my script is based on an example seen on the Liquid Soap workshop 1.0.

I have no issues with Liquidsoap 2.0 alone, but with Stereo Tool it fails most of the time directly with some "PANIC" error messages, or it just hangs or skips to the blank() fallback every second until the program suddenly exits.

I'm wondering if anyone has successfully used the same kind of setup, and if so, would share some advices on how to optimize the whole thing ? For example, I'm not sure wether I should set jackd to RT priority or not...  Also , should I try to use a RT kernel on the RPI ? Perhaps an install on top of a Patchbox OS ? Or do you think the Raspi 3B+ is just too weak to run Liquid Soap , process sound through Stereo Tool, encode, decode, manage Beets, etc at the same time ? I've not started to stream yet, I'm only testing with a direct output to Alsa.

I know there are other really cool ways to process sound within liq or using FFmpeg or Ladspa, but for the sake of simplicity , I would have liked to start with some presets. Just wondering if the Pi3B is a viable solution for that.

Thanks for any help :)

Hardware and software details:
- Rpi 3B+
- Rasbian OS Buster 32bits 
- WM8731 DAC
- Liquidsoap 2.0
- StereoTool 9.23

Here's the script I'm using:

```
#!/usr/bin/env liquidsoap

BEET = "/home/pi/.local/bin/beet"

def beets(arg="") =
  fun() -> list.map(fun(item) -> request.create(item),
     process.read.lines(
      "#{BEET} random -f '$path' #{arg}"
    )
  )
end

def transition(a,b)
  # If old or new source is not music, no fade
  if a.metadata["source_tag"] != "music" or a.metadata["source_tag"] != "music" then
    sequence([a.source, b.source])
  else
    # Else, apply the standard smart transition
    cross.smart(a, b)
  end
end

# A function to add a source_tag metadata to a source:
def source_tag(s,tag) =
  def f(_)
    [("source_tag",(tag:string))]
  end
  map_metadata(id=tag,insert_missing=true,f,s)
end

# Tag our sources
default = blank()
music = request.dynamic.list(beets("length:..03:30"))
music = source_tag(music, "music")
music = amplify(1.,override="replaygain_track_gain", music)
music = skip_blank(music, max_blank=5., threshold=-50.)

radio = music
radio = cross(duration=5., transition, radio)

st = pipe(process='/usr/bin/stereo_tool - - -q', radio)

s = fallback(track_sensitive=false, [st, default])
output.alsa(s, device="pcm.liquidsoap")
```

And this is the kind of output I get from the terminal:

```
[mp3 @ 0x6758b2a0] Estimating duration from bitrate, this may be inaccurate
2021/07/28 19:29:58 [request.dynamic.list_0:3] Prepared "/home/pi/anda/music/jazz/Roland Kirk - Blacknuss - 1972/01 Ain't No Sunshine.mp3" (RID 0).
2021/07/28 19:29:59 [request.dynamic.list_0:2] #get_frame didn't add exactly one break!
2021/07/28 19:29:59 [threads:2] Queue generic queue #1 crashed with exception File "source.ml", line 746, characters 12-18: Assertion failed
pi@radio:~ $ ./radio.liq
no more csLADSPA plugins
2021/07/28 19:36:41 >>> LOG START
2021/07/28 19:36:23 [main:3] Liquidsoap 2.0.0+git@a697602c
2021/07/28 19:36:23 [main:3] Using: bytes=[distributed with OCaml 4.02 or above] pcre=7.5.0 sedlex=2.3 menhirLib=20210419 curl=0.9.1 dtools=0.4.4 duppy=0.9.2 cry=0.6.5 mm=0.7.1 xmlplaylist=0.1.5 ogg=0.7.0 ogg.decoder=0.7.0 vorbis=0.8.0 vorbis.decoder=0.8.0 opus=0.2.0 opus.decoder=0.2.0 mad=0.5.0 dynlink=[distributed with Ocaml] lame=0.3.4 gstreamer=0.3.1 ffmpeg-avutil=97c1fc4 ffmpeg-avcodec=97c1fc4 ffmpeg-avdevice=97c1fc4 ffmpeg-av=97c1fc4 ffmpeg-avfilter=97c1fc4 ffmpeg-swresample=97c1fc4 ffmpeg-swscale=97c1fc4 bjack=0.1.6 alsa=0.3.0 ao=0.2.3 samplerate=0.1.5 taglib=0.3.6 ssl=0.5.9 camomile=1.0.2 faad=0.5.0 ladspa=0.2.0
2021/07/28 19:36:23 [main:3] 
2021/07/28 19:36:23 [main:3] DISCLAIMER: This version of Liquidsoap has been
2021/07/28 19:36:23 [main:3] compiled from a snapshot of the development code.
2021/07/28 19:36:23 [main:3] As such, it should not be used in production
2021/07/28 19:36:23 [main:3] unless you know what you are doing!
2021/07/28 19:36:23 [main:3] 
2021/07/28 19:36:23 [main:3] We are, however, very interested in any feedback
2021/07/28 19:36:23 [main:3] about our development code and committed to fix
2021/07/28 19:36:23 [main:3] issues as soon as possible.
2021/07/28 19:36:23 [main:3] 
2021/07/28 19:36:23 [main:3] If you are interested in collaborating to
2021/07/28 19:36:23 [main:3] the development of Liquidsoap, feel free to
2021/07/28 19:36:23 [main:3] drop us a mail at <savonet-devl@lists.sf.net>
2021/07/28 19:36:23 [main:3] or to join the slack chat at <http://slack.liquidsoap.info>.
2021/07/28 19:36:23 [main:3] 
2021/07/28 19:36:23 [main:3] Please send any bug report or feature request
2021/07/28 19:36:23 [main:3] at <https://github.com/savonet/liquidsoap/issues>.
2021/07/28 19:36:23 [main:3] 
2021/07/28 19:36:23 [main:3] We hope you enjoy this snapshot build of Liquidsoap!
2021/07/28 19:36:23 [main:3] 
2021/07/28 19:36:39 [dynamic.loader:3] Could not find dynamic module for fdkaac encoder.
2021/07/28 19:36:39 [gstreamer.loader:3] Loaded GStreamer 1.14.4 0
2021/07/28 19:36:39 [clock:3] Using builtin (low-precision) implementation for latency control
2021/07/28 19:36:41 [lang.deprecated:2] WARNING: "skip_blank" is deprecated and will be removed in future version. Please use "blank.skip" instead.
2021/07/28 19:36:41 [frame:3] Using 44100Hz audio, 25Hz video, 44100Hz main.
2021/07/28 19:36:41 [frame:3] Video frame size set to: 1280x720
2021/07/28 19:36:41 [frame:3] Frame size must be a multiple of 1764 ticks = 1764 audio samples = 1 video samples.
2021/07/28 19:36:41 [frame:3] Targeting 'frame.duration': 0.04s = 1764 audio samples = 1764 ticks.
2021/07/28 19:36:41 [frame:3] Frames last 0.04s = 1764 audio samples = 1 video samples = 1764 ticks.
2021/07/28 19:36:41 [sandbox:3] Sandboxing disabled
2021/07/28 19:36:41 [video.converter:3] Using preferred video converter: ffmpeg.
2021/07/28 19:36:41 [audio.converter:3] Using samplerate converter: ffmpeg.
2021/07/28 19:36:41 [clock.alsa:3] Streaming loop starts in auto-sync mode
2021/07/28 19:36:41 [clock.alsa:3] Delegating synchronisation to CPU clock
2021/07/28 19:36:41 [switch_0:3] Switch to blank_1.
2021/07/28 19:36:41 [alsa_out(pcm.liquidsoap):3] Using ALSA 1.1.8.
2021/07/28 19:36:41 [alsa_out(pcm.liquidsoap):3] channels=2, samplerate=44100Hz, buffer size=60211B, frame size=4B, periods=2077
2021/07/28 19:36:42 [clock.alsa:3] Delegating synchronisation to active sources
2021/07/28 19:36:52 [request.dynamic.list_0:3] Prepared "/home/pi/anda/music/Various Artists - Strange Breaks & Mr Thing Vol. 2 - 2008/15 Hot City Bump Band - It's Just Begun.mp3" (RID 0).
[mp3float @ 0x67512000] Could not update timestamps for skipped samples.
2021/07/28 19:36:57 [switch_0:3] Switch to pipe_0 with transition.
2021/07/28 19:37:02 [switch_0:3] Switch to blank_1 with transition.
2021/07/28 19:37:04 [switch_0:3] Switch to pipe_0 with transition.
2021/07/28 19:37:06 [switch_0:3] Switch to blank_1 with transition.
2021/07/28 19:37:07 [switch_0:3] Switch to pipe_0 with transition.
2021/07/28 19:37:11 [switch_0:3] Switch to blank_1 with transition.
2021/07/28 19:37:13 [switch_0:3] Switch to pipe_0 with transition.
2021/07/28 19:37:18 [blank.detect_0:2] #get_frame didn't add exactly one break!
2021/07/28 19:37:18 [threads:2] Queue non-blocking queue #2 crashed with exception File "source.ml", line 746, characters 12-18: Assertion failed
2021/07/28 19:37:18 [threads:2] Raised at Process_handler.run.handler in file "tools/process_handler.ml", line 299, characters 16-23
2021/07/28 19:37:18 [threads:2] Called from Duppy.Task.t_of_task.(fun) in file "src/duppy.ml", line 148, characters 50-66
2021/07/28 19:37:18 [threads:2] Called from Duppy.exec in file "src/duppy.ml", line 275, characters 12-21
2021/07/28 19:37:18 [threads:2] Called from Duppy.queue.run in file "src/duppy.ml", line 307, characters 7-24
2021/07/28 19:37:18 [threads:2] Called from Duppy.queue in file "src/duppy.ml", line 357, characters 8-12
2021/07/28 19:37:18 [threads:2] Re-raised at Duppy.queue in file "src/duppy.ml", line 361, characters 8-17
2021/07/28 19:37:18 [threads:2] Called from Tutils.create.(fun).process in file "tools/tutils.ml", line 173, characters 12-15
2021/07/28 19:37:18 [threads:2] 
2021/07/28 19:37:18 [threads:1] PANIC: Liquidsoap has crashed, exiting.,
2021/07/28 19:37:18 [threads:1] Please report at: savonet-users@lists.sf.net
pi@radio:~ $ 
```

**Comment by toots** [comment, marked as answer]
Hi! This appears to be a bug so I'm converting this as one, thanks for reporting.

Regarding `pip` with `stereotool`, I believe most users do disable the mechanism that closes and restarts the process on each track and adjust the `replay_delay` parameter. 

I just updated the doc for it:
```
 * replay_delay : float? (default: null)
     Replay track marks and metadata from the input source on the output
     after a given delay. If `null` (default) close and flush the process on
     each track and metadata to get an exact timing. This parameter is
     typically used when integrating with `stereotool`.
```

**Reply by nordseele** [reply]
Alright, thanks Toots ! :)
I'll work on the rest of my script and I'll leave the output processing for later until this gets fixed. Or I'll simply use the Ladspa or FFmpeg treatment in the meantime. 
Yes, I found the `replay_delay` in one of the examples of the book after posting this question. I'll read the documentation in more depth.

Thanks again.

**Reply by rrolla** [reply]
> Hi! This appears to be a bug so I'm converting this as one, thanks for reporting.
> 
> Regarding `pip` with `stereotool`, I believe most users do disable the mechanism that closes and restarts the process on each track and adjust the `replay_delay` parameter.
> 
> I just updated the doc for it:
> 
> ```
>  * replay_delay : float? (default: null)
>      Replay track marks and metadata from the input source on the output
>      after a given delay. If `null` (default) close and flush the process on
>      each track and metadata to get an exact timing. This parameter is
>      typically used when integrating with `stereotool`.
> ```

This is something new to me, before there was advice to use merge_tracks https://github.com/savonet/liquidsoap/issues/885#issuecomment-516661378

**Reply by toots** [reply]
Ha I see. Yes, both advice relate to avoiding a restart of the process I believe.

**Comment by KevanGP** [comment]
Does anyone know how to get rid of the 5 second delay I'm getting when Stereo Tool is on through Liquidsoap?
I asked on the Stereo Tool forums, and they say it's Liquidsoap's doing.

**Reply by toots** [reply]
Hi. I'm not sure but one thing we've been discussing is using stereotool VST plugin (or LADSPA). Would you have any interest asking the author(s) if they'd be interested in providing a shared library for linux (preferable LADSPA because we already have it but VST could do)? This way we could integrate stereotool directly as an operator instead of having to go through an external process..

**Reply by hansvanzutphen** [reply]
Hi, Hans from Thimeo Audio Technology (the maker of Stereo Tool) here. Do you actually mean LADSPA, or LV2? When I search for LADSPA I can't find anything that's less than 19 years old.

I haven't looked deep enough yet, but if we do make an LADSPA/LV2 plugin, do you have a way to load settings files and setting a license key? We have more than 8000 parameters, so offering them in the way that most plugins appear to do it (as LADSPA or LV2 parameters) isn't going to be very usable. So I think we would basically need to have an interface with 2 parameters: The preset file and a license key.

We can also provide a VST plugin; before Steinberg didn't support Linux but I just checked and that appears to have changed. I assume that you would use VST3? Note that using VST plugins without having access to the GUI might be difficult to use and might require that you write non-standard code to import presets. VST plugins offer multiple ways of storing parameters; the problem is that as far as I know (I could be wrong) it's not defined how a host should store them, so there's no standardized way to import settings that have been created elsewhere.

We also have our own plugin standard, which is really simple to integrate (4 calls suffice - 5 if you want to load a preset file, and the license key is one of the parameters), but of course it won't help with supporting other software.

**Reply by toots** [reply]
Hi @hansvanzutphen and thanks for chiming in! If y'all already have an existing shared object we can build upon I think we could definitely directly use it. Is there any doc/link I could look at? I was looking at the download page but didn't find any info on it.

**Reply by hansvanzutphen** [reply]
Hi, you can find it here: https://download.thimeo.com/Stereo_Tool_Generic_plugin.zip . This contains the header files and dynamic libraries for Windows, Linux (Intel/AMD and ARM), Mac and iOS.

The file Generic_StereoTool.h contains a description of the functions and a short example program (also pasted below).

The following example program loads the plugin, loads a preset, turns the equalizer on, sends a tone through it (expected sample values are typically between -32768 and +32767 instead of -1 and +1; you can exceed that without issues, it won't clip) and unloads the plugin. You can load multiple instances if needed (at most 128 at the moment; we can increase that if needed).

```
gStereoTool* st = stereoTool_Create(NULL);
bool r = stereoTool_LoadPreset(st, "C:\\temp\\generic_stereo_tool.sts", ID_SAVE_ALLSETTINGS);
assert(r); // stop if loading preset failed
stereoTool_SetStsValue(st, PARAM_equalizer_on, 0, "1");
float smp[1024 * 2];
float sin_l = 0;
float sin_r = 0;
for (int i=0; i<100; i++)
{
    for (int s=0; s<1024; s++)
    {
        smp[2 * s    ] = sin(sin_l) * 32768.0f;
        smp[2 * s + 1] = sin(sin_r) * 32768.0f;
        sin_l += .1f;
        sin_r += .11f;
    }
    stereoTool_Process(st, smp, 1024, 2, 44100);
}
stereoTool_Delete(st);
```

**Reply by toots** [reply]
Awesome thanks, I'll have a look shortly.

**Reply by toots** [reply]
PR implementing the operator is here: https://github.com/savonet/liquidsoap/pull/2953 !

**Reply by hansvanzutphen** [reply]
Great! I assume that this means that I should add the .so file to our download page? So far, others who have integrated it are just including the .so with their download (that's why it wasn't available directly on he site), but I'm guessing that you don't want to include proprietery binaries in your downloads.

**Reply by toots** [reply]
Yes, that would be ideal, thanks!

While we're talking, I am not doing much error around license. What happens if the API is used without a license key? 

We do test for invalid license after making the latency call which, I believe, does some audio processing but we are not raising any error if the result is `false`.

Any error I should be checkin on and reporting back?

**Reply by hansvanzutphen** [reply]
Without a license or with an invalid license, it will add spoken texts and/or beeps in the audio. Everything will work, so it's available for testing, but not really usable.

You can call these functions:
stereoTool_CheckLicenseValid (I assume you're already calling this one)
stereoTool_GetUnlicensedUsedFeatures

to verify if a license is valid, indeed after passing some audio through it. The 2nd function returns a boolean and a text string with a list of features that are used without a license, which you could display. Example output: "Declipper used without license".

**Reply by toots** [reply]
Ok great! In this case we won't raise an error and prevent the script from running and leave this to the user to check valid output and logs. Thanks!

**Reply by hansvanzutphen** [reply]
When I add the .so files to our site (which ones should I add? 64 bit Intel/AMD Linux, 32 and 64 bit ARM, 32 and 64 bit Windows, macOS, iOS, ...?), do you have a description of how to use it that I can link to? Or should I write my own?

**Reply by toots** [reply]
I will write some doc and blog post soon. The operator has a parameter to point to the shared library, another for the preset file and one for the license key. I'll have that shortly and will report. We provide binaries for the following platforms currently:
* amd64, linux/macos
* arm64, linux/macos
* windows 64 bit

arm 32 bit is currently provided but it will most likely be deprecated for the release that will have support for the operator.

Thanks for your help with this!

**Reply by toots** [reply]
Ok doc is here: https://www.liquidsoap.info/doc-dev/stereotool.html. I'm wrapping up the builds for the latest rolling release and it'll be good on my end. Let me know if you spot any issue with the content.

**Reply by toots** [reply]
Release assets now support the operator at: https://github.com/savonet/liquidsoap/releases/tag/rolling-release-v2.2.x

**Reply by hansvanzutphen** [reply]
I've added a link to the SDK .zip file on our download page with links to your doc, for the next software release we'll upload and link the separate .so, .dll and .dylib files. https://www.thimeo.com/stereo-tool/download/

One final question: For Linux, do people typically use Liquidsoap on a system with no X11 support? We have libraries with and without X11 support, and I can link either. In most other programs, people will use the version with X11 support so they can open the GUI, so it's a bit easier to only supply those - but on a system with no X11 libraries installed at all that won't run.

**Reply by toots** [reply]
The software is mostly for servers with no graphical interface so a build without X11 would be ideal. The only functions used in the operator currently are:
```
stereoTool_CheckLicenseValid
stereoTool_Create
stereoTool_Delete
stereoTool_GetApiVersion
stereoTool_GetLatency2
stereoTool_GetSoftwareVersion
stereoTool_GetUnlicensedUsedFeatures
stereoTool_LoadPreset
stereoTool_Process
stereoTool_Reset
stereoTool_SavePreset
```

**Reply by DroidU** [reply]
Could we ask for a concrete example?
There are many people who use Liquidsoap for a simple target task. If I could try a concrete example, it would be immediately clear. For example, processing an incoming http stream with Stereo Tool and then transcoding it to another Icecast2 server. All this in such a way that the metadata also works when changing songs.

**Reply by toots** [reply]
There's a documentation here: https://www.liquidsoap.info/doc-dev/stereotool.html and also on lthe blog: https://www.liquidsoap.info/blog/2023-03-16-stereotool-shared-library-support/

The operator should be really be transparent and easy to use so, for a simple setup:

```liquisoap
s = input.http(...)

s = stereotool(
  library_file="/path/to/stereotool/shared/lib",
  license_key="my_license_key",
  preset="/path/to/preset/file",
  s
)

output.icecast(..., s)
```

**Reply by DroidU** [reply]
I have managed to get it to work without giving an error message. Sending the stream to Icecast2 also works, but there is only silence.
If I comment/delete out the stereotool parameters, so I start it without them, it works perfectly.

Config: https://pastebin.com/H8RY5Fbi
Log: https://pastebin.com/aFL8sGwP

**Reply by toots** [reply]
Can you send me your parameters file so I can test it locally? toots@rastageeks.org

**Reply by DroidU** [reply]
I sent it.

**Reply by toots** [reply]
Your script works here. What OS/arch are you using?

**Reply by DroidU** [reply]
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

The only changes are that the deb-multimedia repository is included: "deb http://www.deb-multimedia.org bullseye main non-free"

**Reply by toots** [reply]
Architecture?

**Reply by DroidU** [reply]
x86_64, and the test environment runs in a VirtualBox.

**Reply by toots** [reply]
This works on amd64 debian bookworm (our production docker image):

```liquidsoap
s = input.http("https://icecast.radiofrance.fr/fip-hifi.aac")

s = stereotool(
  library_file="/tmp/Stereo_Tool_Generic_plugin_1000/libStereoTool_64.so",
  preset="/tmp/test.sts",
  s
)

output.icecast(
  mount="test",
  password="hackme",
  port=8000,
  host="192.168.86.206",
  %mp3,
  s
)
```

**Reply by toots** [reply]
Your script works but the it appears that you need a license for some of the features you are using. This was actually shown in the log:
```
2023/04/06 18:53:27 [stereotool_0:4] Stereotool initialized! Valid license: false, latency: 0.09s, API/software version: 256/10000
2023/04/06 18:53:27 [stereotool_0:4] Using unlicensed features: Advanced Clipper used without license
```
The output glitch is fixed in the latest code now

**Reply by toots** [reply]
Hi @hansvanzutphen. What's the preferred channel for reports? I've noticed two minor glitches:
* When `stereoTool_CheckLicenseValid` is `false, and making two calls to `stereoTool_GetUnlicensedUsedFeatures`, the second one returns `false`. It would good if the returned string could be idempotent and always consistent with the result of `stereoTool_CheckLicenseValid`.
* On macos, calls to `stereoTool_GetUnlicensedUsedFeatures` never return anything even when `stereoTool_CheckLicenseValid` is `false`.

**Reply by DroidU** [reply]
I tried it with a working but not full-feature license key, and I also tried it by deleting the license_key variable. It fails in both cases. I also took a closer look at the Icecast2 server log.
**[2023-04-07  08:37:26] WARN source/get_next_buffer Disconnecting source due to socket timeout**
This disconnection is not logged by Liquidsoap. So it seems that things get stuck somewhere, because after connecting, no data goes to Icecast2. And because of this, it drops the connection.
It might be due to Debian 11. But even Debian 10 is used by a lot of people.
If the bug persists in a newer Liquidsoap version, I may send the VirtualBox vdi file to the above email address.

**Reply by toots** [reply]
Thanks for pushing back. I tracked it down to a bug in the new muxer code. Should be fixed in `main` and `rolling-release-v2.2.x` now (both are currently building)

**Reply by DroidU** [reply]
I also experimented with encoding to mp2 (twolame) stream without stereotool. It works perfectly in stereo, but if I encode the same in mono, it stops/gets stuck in the same way as in the case of stereotool.
**output.icecast(%ffmpeg(format="mp2",%audio(codec="mp2",samplerate=44100,b="128k",ac=1)),host="127.0.0.1",port=8000,mount="/trans.mp3",user="source",password="hackme",icy_metadata="true",mean(mksafe(s)))**

CTRL+C to stop doesn't work with stereotool and in this case either.

**Reply by toots** [reply]
I can confirm that this use-case is also fixed now.

**Reply by DroidU** [reply]
I tried 471bd7c.
Stream playback is now near perfect in all cases.
However, the stream converted to ogg/vorbis has a very slight skip when changing songs, with "Unknown" appearing instead of the song title. The tiny skip itself may not occur with all players, but it certainly does with VLC.
Also an ogg problem is that if the stream being processed is ogg, all encoder channels display "Unknown" as the song title. And the OPUS and OggFLAC song title display is blank.

**Comment by moleculezz** [comment]
I'm on x86_64. What files from the Stereo Tool generic plugin do I have to copy to the system?

**Reply by hansvanzutphen** [reply]
Just libStereoTool_64.so.

**Reply by toots** [reply]
Hi @hansvanzutphen . I just sent an email to support@thimeo.com with what seems to be an issue with the dynamic plugin erroneously reporting invalid keys. Is there a better way to report issues? Thanks!
